grafana_client/api.py

```diff
@@ -32,6 +31,28 @@
     User,
     Users,
 )
+from .elements._async import (
+    AsyncAdmin,
+    AsyncAlerting,
+    AsyncAlertingProvisioning,
+    AsyncAnnotations,
+    AsyncDashboard,
+    AsyncDashboardVersions,
+    AsyncDatasource,
+    AsyncFolder,
+    AsyncHealth,
+    AsyncNotifications,
+    AsyncOrganization,
+    AsyncOrganizations,
+    AsyncPlugin,
+    AsyncRbac,
+    AsyncSearch,
+    AsyncServiceAccount,
+    AsyncSnapshots,
+    AsyncTeams,
+    AsyncUser,
+    AsyncUsers,
+)
```

Comment:
Hi Ahmed,

thank you very much for the contribution. As Niquests advertised itself as a drop-in replacement, I actually would have expected that the patch would be smaller, without needing to replicate code.

> With this, should we create a script that automate the generation of the `_async` submodule?

To be honest, I don't like the duplication / redundancy at all. Can you describe the problem why code duplication is necessary?

With kind regards,
Andreas.

Reply:
Hello,

> As Niquests advertised itself as a drop-in replacement, I actually would have expected that the patch would be much smaller, without needing to duplicate code. 

The code duplication has nothing to do with Niquests, it is a Python constraint. To make a code async, you have to use the keyword async/await in the code tree, you can't just return the coroutine in the sync enclosure, won't work.

Actually, the size of the duplication can vary, depending of the code architecture. Unfortunately, there, to ensure the same code, we can't do smaller unless mistaken.

 Here are the scenario accessible:

- Code duplication, no breaking changes
- Reduce the code duplication, breaking changes

Generating the async part is common enough to serve both async and sync code.

For another story, if async is out the equation, you can consider leveraging a multiplexed connection. While it is not async, it is a potential substantial boost.

Regards,

Reply:
Thank you for your quick reply, Ahmed. I think way to go will to make everything async then, and overlay it with a synchronous interface for those who need it, in order to avoid code duplication?

Reply:
We can move this way. If I understand it correctly, you mean to leverage `get_event_loop+run_until_complete` available in asyncio when in a synchronous context? The thing with this, while it will work, can induce a useless complexity in the sync enclosure, consume CPU time, can be unstable if someone is actually using the event loop somewhere else, etc...
I must say that, while it is possible, the code duplication still offers better performances and less cpu pressure (sync mode).
WDYT?

Reply:
Thank you for elaborating on this. Are there any guidelines about the details you are referring to, or other blueprints where projects are doing it in the same way? If so, there may already be corresponding code transformation tools around?

Reply:
>  where projects are doing it in the same way?

Of course, immediately, in mind I have huggingface https://github.com/huggingface/huggingface_hub/blob/main/utils/generate_async_inference_client.py There's a lot similar to this, but I don't have them in mind right away.

>  If so, there may already be corresponding code transformation tools around?

Some heuristic tool exist, but they often require full typing and some configuration,
For this project, I have just composed a script that automatically generate the async counterpart code.

Please find it in `script/generate_async.py`, and I have added some docs to warn contributors about it.

> we would also need to transform all the testing code correspondingly?

Fortunately, no, we don't. 
My initial move on this would to use tool like mypy to verify that the code is correct and we didn't miss anything. But the project current state is untyped for the time being.

Finally I have configured coverage to exclude anything in `_async/*`.

Regards,

Reply:
is there anything else I can do or answer?

Reply:
Hi again. Apologies that we haven't been able to follow up on this in a timely manner. Regarding tool support, to only maintain a single code base, I've just discovered those programs. Would they be of any help in this regard, if they work well?

- https://github.com/python-trio/unasync
- https://github.com/provinzkraut/unasyncd

Reply:
Hello again,

> Regarding tool support, to only maintain a single code base, I've just discovered those programs. Would they be of any help in this regard, if they work well?

They would help, but the result will be the same, ultimately the code will be in git. Integrating one of these tool, preferably unasyncd, would require quite some time. With that approach we would have to test the async part first, because testing a generated piece of code is insufficient to guarantee the project stability (as the user base is exclusive to sync).

`generate_async.py` doesn't handle the client itself but rather the repetitive and safe part that can be converted with ease. I think that this PR is a "good" starting point, then followup PRs could improve the package by safely experimenting reversing the generation logic.

regards,

Reply:
gentle ping

Reply:
Hi,

I understand, the priorities are what they are.
So, great, I think everything is ready.

I am onboard with the proposed solution.

> Do you have any objections about this procedure, or different suggestions?

None. Let's do it.